spec/unit/weberSsim.spec.ts

```diff
@@ -0,0 +1,184 @@
+import { Options } from "../../src/types";
+import { average, covariance, mean2d, variance } from "../../src/math";
+import { defaults } from "../../src/defaults";
+import { ssim } from "../../src/index";
+import { roundTo } from "../helpers/round";
+import { sampleCsv, samples } from "../helpers/matrices";
+import {
+  partialSumMatrix1,
+  partialSumMatrix1Forward,
+  windowCovariance, windowCovarianceForward, windowMatrix, windowMatrixForward,
+  windowSums,
+  windowVariance, windowVarianceForward
+} from "../../src/weberSsim";
+import { sub } from "../../src/matlab";
+
+
+const testDataImg = {
+  width: 4, height: 4, data: [
+    0x10, 0x20, 0x30, 0x40,
+    0x50, 0x60, 0x70, 0x80,
+    0x90, 0xa0, 0xb0, 0xc0,
+    0xd0, 0xe0, 0xf0, 0xff
+  ]
+};
+const testDataImg2 = {
+  width: 4, height: 4, data: [
+    0x0c, 0x1c, 0x2c, 0x3c,
+    0x4c, 0x5c, 0x6c, 0x7c,
+    0x8c, 0x9c, 0xac, 0xbc,
+    0xcc, 0xdc, 0xec, 0xfc
+  ]
+};
+
+const testPSForwardImg1 = {
+  width: 5, height: 5, data: new Int32Array([
+    0, 0, 0, 0, 0,
+    0, 16, 48, 96, 160,
+    0, 96, 224, 384, 576,
+    0, 240, 528, 864, 1248,
+    0, 448, 960, 1536, 2175])
+};
+
+
+describe("weberSsim", () => {
+  let options: Options;
+  let k0Options: Options;
+  let optionsForward: Options;
+  let k0OptionsForward: Options;
+
+  beforeEach(() => {
+    options = { ...defaults, ssim: "weber" };
+    k0Options = { ...defaults, ssim: "weber", k1: 0, k2: 0 };
+    optionsForward = { ...defaults, ssim: "weberForward" };
+    k0OptionsForward = { ...defaults, ssim: "weberForward", k1: 0, k2: 0  };
```

Comment:
I didn't realize before, why do have 2 formats?

Reply:
It's new, actually.  I'm trying to optimize the implementation so that it works forwards and backward.  The original uses backwards memory accesses.  I wanted to build a forward implementation so I can then vectorize it.  Technically the forward implementation should be faster since it work's like the CPU thinks it should, but I can't tell which is because javascript does different optimizations....

Reply:
yeah not sure why that happens in JS, in my experience most browsers handle decremental while loops faster than forward for loops but you would expect more cache misses that way.

Sounds like a separate exploration in progress. What if we keep just the `weber` format and once we have an improvement on it, we replace it. I want to avoid having versioning inside the library (weber, weberForward, weber2.0, weberFinalFinal... 😉 )

Reply:
I understand the sentiment...  I just don't want to throw away the work, and now they test against each other.

Reply:
I made a new branch to store the forward implementation.  Will push the slimmed down version shortly.